### PR TITLE
Integrating wagmi and web3auth

### PR DESCRIPTION
**Pull request by fabriziogianni7**
# Description

Integrating [Wagmi hooks](https://wagmi.sh/core) in botblock.

It will just replace The previous web3Context. We will discard the web3 context and we will push on wagmi. NVM app uses wagmi so ideally, integrating nvm in botblock will be easier

## Components to have a look at
- **useBBContractReads**: hook to read multiple contracts when invoked
- **useBBContractWrite**: hook to invoke functions of contracts
- **useBBMulticall**: hook to invoke more contract calls at once

## Notes
Purchasing the plan is currently not working - multicall can't work for approval and spending of the token, but since the botblock contract will be replaced by the NVM marketplace, we can ignore this problem for now

### CODE REVIEW

File: packages/nextjs/pages/profile/index.tsx [hunk 1]
```diff
@@ -1,34 +1,35 @@
...
     };
 
-    if (subsContract && address) {
-      fetchPlans();
-    }
-  }, [address, getPlans, subsContract]);
+
+    fetchPlans();
+
+  }, [isConnected]);
```
Comment (juanmanso) [inline packages/nextjs/pages/profile/index.tsx:24]: should we wait for the `allPlans` to resolve? (I'm not sure if it is fetched by the `useBBContractReads` hook)

If so, we should add an if statement and add it as a dependency

File: packages/nextjs/pages/profile/index.tsx [hunk 2]
```diff
@@ -1,34 +1,35 @@
...
-      setUserOrders(orders?.filter(order => order.buyer === address));
+      setUserOrders(allOrders?.filter(order => order.buyer === address));
     };
+    fetchOrders();
 
-    if (subsContract && address) {
-      fetchOrders();
-    }
-  }, [address, getOrders, subsContract]);
+  }, [isConnected]);
```
Comment (juanmanso) [inline packages/nextjs/pages/profile/index.tsx:32]: same here and if we are not awaiting anything, we could remove the wrapper completely.

File: packages/nextjs/pages/subscribe/[planId].tsx [hunk 3]
```diff
@@ -83,19 +110,19 @@ const SubscriptionDetails = () => {
 
   useEffect(() => {
     const setPlanIfExists = async () => {
-      const plans = await getPlans();
-      const foundPlan = plans?.find(plan => plan.planId === planId);
+      // const plans = await getPlans();
```
Comment (juanmanso) [inline packages/nextjs/pages/subscribe/[planId].tsx:113]: nitpick: commented code can be removed here right?

**Review by juanmanso — APPROVED** [review]
Sorry for the late review, +400 lines took me some time haha

It's a big PR and did not find any bugs. There might be stuff that could be a code smell and I think I pointed it out on my comments, let's chat about them and I'm open to doing these changes on a follow-up PR. Nonetheless, great work!

Leaving the approve for future merging

File: packages/nextjs/components/Header/index.tsx [hunk 4]
```diff
@@ -114,8 +113,7 @@ export const Header = () => {
         <ul className="hidden lg:flex lg:flex-nowrap menu menu-horizontal px-1 gap-2">{navLinks}</ul>
       </div>
       <div className="navbar-end flex-grow mr-4">
-        <LoginButton />
-        <FaucetButton />
```
[the remark this answers is not in the document]
Comment (fabriziogianni7) [reply packages/nextjs/components/Header/index.tsx:118]: yeah, it is true. need to be replaced over the app

File: packages/nextjs/context/Types.tsx [hunk 5]
```diff
@@ -1,5 +1,5 @@
 export type Plan = {
-  planId?: number | string;
+  planID?: number | string;
```
[the remark this answers is not in the document]
Comment (fabriziogianni7) [reply packages/nextjs/context/Types.tsx:2]: yeah, the smart contract returns the plan object with planID

File: packages/nextjs/components/Header/index.tsx [hunk 6]
```diff
@@ -2,15 +2,14 @@ import React, { useCallback, useRef, useState } from "react";
 import Image from "next/image";
 import Link from "next/link";
 import { useRouter } from "next/router";
-import { LoginButton } from "./LoginButton";
```
[the remark this answers is not in the document]
Comment (fabriziogianni7) [reply packages/nextjs/components/Header/index.tsx:5]: it is replaced by  <RainbowKitCustomConnectButton />

[the remark this answers is not in the document]
Comment (fabriziogianni7) [reply packages/nextjs/components/Header/index.tsx:5]: `RainbowKitCustomConnectButton`

File: packages/nextjs/hooks/Botblock/useBBContractWrite.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import { useContractWrite, usePrepareContractWrite } from "wagmi";
+import { notification } from "~~/utils/scaffold-eth";
+
+/**
+ * @param contractName: name of the contract (needed to load abi)
+ * @param functionName: name of the function
+ */
+
+export function useBBContractWrite({ contractName, functionName }: UseBBContractWrite) {
+  // address?: string
```
Comment (juanmanso) [inline packages/nextjs/hooks/Botblock/useBBContractWrite.ts:12]: nitpick: commented code

**Review by juanmanso — APPROVED** [review]
Great job man, left some small nitpicks and comments but thanks for applying the changes 😄

File: packages/nextjs/services/web3/rainbowWeb3AuthConnector.tsx [hunk 8]
```diff
@@ -24,7 +34,7 @@ export const rainbowWeb3AuthConnector = ({ chains }) => {
 
   const web3AuthInstance = new Web3AuthNoModal({
     clientId: CLIENT_ID as string,
-    chainConfig: chainConfig as any,
+    chainConfig: chainConfig as ChainConfig,
```
[the remark this answers is not in the document]
Comment (fabriziogianni7) [reply packages/nextjs/services/web3/rainbowWeb3AuthConnector.tsx:37]: This was giving error without casting it

[the remark this answers is not in the document]
Comment (juanmanso) [reply packages/nextjs/services/web3/rainbowWeb3AuthConnector.tsx:37]: Okay, was thinking of a change like this actually 😅 

image

**Pull request merged**